raft/raft.go

```diff
@@ -1106,32 +1106,93 @@ func stepLeader(r *raft, m pb.Message) error {
+			// which can easily result in hours of time spent probing and can
+			// even cause outright outages. The probes are thus optimized as
+			// described below.
+			r.logger.Debugf("%x received MsgAppResp(rejected, hint: (index %d, term %d)) from %x for index %d",
+				r.id, m.RejectHint, m.LogTerm, m.From, m.Index)
+			nextProbeIdx := m.RejectHint
 			if m.LogTerm > 0 {
-				// LogTerm is the term that the follower has at index RejectHint. If the
-				// follower has an uncommitted log tail, we would end up probing one by
-				// one until we hit the common prefix.
+				// If the follower has an uncommitted log tail, we would end up
+				// probing one by one until we hit the common prefix.
 				//
 				// For example, if the leader has:
-				//   [...] (idx=1,term=2) (idx=2,term=5)[...](idx=9,term=5)
-				// and the follower has:
-				//   [...] (idx=1,term=2) (idx=2,term=4)[...](idx=6,term=4)
 				//
-				// Then, after sending (idx=9,term=5) we would receive a RejectHint of 6
-				// and LogTerm of 4. Without the code below, we would try an append at
-				// 6, 5, 4 ... until hitting idx=1 which succeeds (as 1 matches).
-				// Instead, the code below skips all indexes at terms >4, and reduces
-				// 'hint' to 1 in one term (meaning we'll successfully append 2 next
-				// time).
-				hint = r.raftLog.findConflictByTerm(hint, m.LogTerm)
+				//   idx        1 2 3 4 5 6 7 8 9
+				//              -----------------
+				//   term (L)   1 3 3 3 5 5 5 5 5
+				//   term (F)   1 1 1 1 2 2
+				//
+				// Then, after sending an append anchored at (idx=9,term=5) we
+				// would receive a RejectHint of 6 and LogTerm of 2. Without the
+				// code below, we would try an append at index 6, which would
+				// fail again.
+				//
+				// However, looking only at what the leader knows about its own
+				// log and the rejection hint, it is clear that a probe at index
+				// 6, 5, 4, 3, and 2 must fail as well:
+				//
+				// For all of these indexes, the leader's log term is larger than
+				// the rejection's log term. If a probe at one of these indexes
+				// succeeded, its log term at that index would match the leader's,
+				// i.e. 3 or 5 in this example. But the follower already told the
+				// leader that it is still at term 2 at index 9, and since the
+				// log term only ever goes up (within a log), this is a contradiction.
+				//
+				// At index 1, however, the leader can draw no such conclusion,
+				// as its term 1 is not larger than the term 2 from the
+				// follower's rejection. We thus probe at 1, which will succeed
+				// in this example. In general, with this approach we probe at
+				// most once per term found in the leader's log.
+				//
+				// There is a similar mechanism on the follower (implemented in
+				// handleAppendEntries via a call to findConflictByTerm) that is
+				// useful if the follower has a large divergent uncommitted log
+				// tail, as in this example:
```

Comment:
Consider mentioning that this is only useful if the follower's large divergent tail has terms that are larger than those at the corresponding indexes in the leader's log. In the process of understanding this issue, I've had to convince myself of that at least three times. I've also had to convince myself that this situation is even possible, which it is.